### DISCUSSION TITLE
SPEC 1 — Lazy Loading of Submodules and Functions

### DISCUSSION THREAD

**Discussion by tkoyama010** [pyvista/pyvista#5512] [Ideas]
I'd be really interested to know your thoughts on [lazy_loader](https://github.com/scientific-python/lazy_loader/) and whether it's appropriate to be adopted in `pyvista` ?

Also see [SPEC 1 — Lazy Loading of Submodules and Functions](https://scientific-python.org/specs/spec-0001/).

_Originally posted by @bjlittle in https://github.com/bjlittle/geovista/issues/669#issuecomment-1909191228_

**Comment by bjlittle** [comment]
Just for some context, I was talking to the devs across on `mne-python` about `lazy_loader`, and their journey towards adoption was interesting, see this [comment](https://github.com/mne-tools/mne-python/issues/12388#issuecomment-1910876635).

Anyways, I've adopted `lazy_loader` across on [geovista](https://github.com/bjlittle/geovista) as it helped us resolve this issue https://github.com/bjlittle/geovista/issues/647 regarding a sluggish `import geovista`.

In a nutshell, we've gone from ~2.5secs to >100msecs.

I don't know the history behind streamlining the `import` mechanism of `pyvista`, but I'm sure it's long, complex and involved ... but I'd love to know more.

I'd be super interested to know whether there is general appetite or not for adopting the `lazy_loader` approach, as ultimately (and selfishly) `geovista` could still benefit further as a result (as with the wider `pyvista` community).

That said, I'd be happy to put in the effort to make this happen, but only if other core devs are in agreement, and also if there's a core dev who I could ask questions, if they arise. However, ultimately I'd do the grunt work to make this cross the line.

**Reply by banesullivan** [reply]
> their journey towards adoption was interesting, see this https://github.com/mne-tools/mne-python/issues/12388#issuecomment-1910876635.

We're already seeing issues with intellisense and other IDE tools with our current lazy loading of the plotting subpackage, ref #4742 and historically have had a hard time getting mypy happy with pyivsta.

> general appetite or not for adopting the lazy_loader approach

My appetite is scarce these days. It is a lot of work to adjust the imports making the gains seem marginal to me given the cost of developer time, broken intellisense and tab-completion for users, issues with type checking/mypy, and the overhead of maintaining tests that validate the import structure. Some relevant issues/PRs include: #4445, #4486, #4629. Further, it adds a lot of overhead for maintainers and new contributors.

In my opinion, the current state of our imports in PyVista strikes a nice balance here. We have two subpackages: 1) `pyvista.core` and 2) `pyvista.plotting`. `import pyvista` will immediately import all of `pyvista.core` for immediate use and has a custom `__getattr__` to lazily import anything from `pyvista.plotting`. The reason for lazily loading `pyvista.plotting` was to keep PyVista's core data model wrappings, file IO, and filters available in environments without any GL libraries. Its my opinion that this seperation provides a clear benefit to many of our users: being able to `pip install pyvista` just about anywhere and use the entire core API.

**Reply by banesullivan** [reply]
To be honest, I just don't get why import time matters? I'm curious what the root of the issue is here and in what setting pyvista is being used where import times are important. Asking because there may be other things we can do to improve performance in those settings specifically.

**Reply by bjlittle** [reply]
Thanks @banesullivan for taking the time to summarise the situation, it's really useful to know.

For me the import time on `geovista` wasn't really on my radar. However, several new users mentioned this to me in passing, but I didn't consider it particularly important or an issue as such.

After spending hands-on time introducing `geovista` to the development team I work with, a few of them also mentioned this to me ... hence reluctantly I put some effort into seeing where time was being spent.

To be honest, I'm in agreement with you in that this would require quite a bit of effort for `pyvista`, for questionable gains, plus I sense that it might introduce more issues than it attempts to resolve.

I'm kinda of the opinion to let sleeping dogs lie here.

Although I've implemented `lazy_loader` on `geovista` the jury is still out. I'm going let the dust settle and see how it behaves.

**Reply by banesullivan** [reply]
> plus I sense that it might introduce more issues than it attempts to resolve.

This is the biggest thing preventing me from getting on board at this time.

> I'm kinda of the opinion to let sleeping dogs lie here.

Well put

> Although I've implemented lazy_loader on geovista the jury is still out. I'm going let the dust settle and see how it behaves.

I think having the lazy loader downstream may go a long way for your users. Still the same import times, just spread out hopefully so its less noticeable. Curious to see how this turns out for geovista users
